# Исследование надежности заемщиков


## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [2]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [15]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [16]:
data.duplicated().sum()

71

In [17]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [18]:
def categorize_income(income):
    try:
        if income <= 30000:
            return 'E'
        elif income <= 50000:
            return 'D'
        elif income <= 200000:
            return 'C'
        elif income <= 1000000:
            return 'B'
        else:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [23]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
# Сначала хотелось бы понять можем ли мы использовать все категории заемщиков (по количеству детей) для оценки и анализа
data.groupby('children')['debt'].count()

children
0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: debt, dtype: int64

In [24]:
# Выведем данные по категориям заемщиков (по количеству детей), у которых были "проблемы" по возврату кредитов 
data.groupby('children')['debt'].sum()

#этот шаг не обязательно выводить отдельно, можно сразу перейти к следующему


children
0    1063
1     444
2     194
3      27
4       4
5       0
Name: debt, dtype: int64

In [25]:
# Следующим шагом увидим % проблемных заемщиков в группе (по количеству детей)
ratio_by_children = data.groupby('children')['debt'].mean()
(ratio_by_children * 100).round(2).astype(float).astype(str) + '%'

children
0    7.54%
1    9.23%
2    9.45%
3    8.18%
4    9.76%
5     0.0%
Name: debt, dtype: object

In [26]:
# те же результаты с помощью pivot_table - в колонке count количество заемщиков в группе, 
# в колонке mean - доля проблемных заемщиков в этой группе
data_pivot_children = data.pivot_table(index='children', values='debt', aggfunc={'count', 'mean'})
data_pivot_children

count      mean
children                   
0         14091.0  0.075438
1          4808.0  0.092346
2          2052.0  0.094542
3           330.0  0.081818
4            41.0  0.097561
5             9.0  0.000000

**Вывод:** 
Группами с 3, 4 и 5 детьми в подготовке выводов можно пренебречь, так как эти группы очень малы относительно общего количества заемщиков(всего ~ 1.8%).
Доля проблемных клиентов значительно меньше в группе без детей. Вероятнее всего это связанно с частыми непредвиденными расходами на нужды ребенка, а также возможно с частыми пропусками работы родителями детей в связи с болезнями. 

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [27]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
# Посмотрим на количество заемщиков в разбивке по семейному положению
data.groupby('family_status')['debt'].count()

family_status
Не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
гражданский брак          4134
женат / замужем          12261
Name: debt, dtype: int64

In [28]:
# Посмотрим на количество заемщиков в разбивке по семейному положению
data.groupby('family_status')['debt'].sum()
#этот шаг не обязательно выводить отдельно, можно сразу перейти к следующему

family_status
Не женат / не замужем    273
в разводе                 84
вдовец / вдова            63
гражданский брак         385
женат / замужем          927
Name: debt, dtype: int64

In [29]:
# Следующим шагом увидим % проблемных заемщиков в группе 
ratio_by_family_status = data.groupby('family_status')['debt'].mean()
(ratio_by_family_status * 100).round(2).astype(float).astype(str) + '%'

family_status
Не женат / не замужем    9.76%
в разводе                7.06%
вдовец / вдова           6.62%
гражданский брак         9.31%
женат / замужем          7.56%
Name: debt, dtype: object

**Вывод:** 
Заемщики, которые были женаты/замужем, причем не важно в каком статусе они сейчас (женат/замужем, в разводе, вдовец/вдова) статистически благонадежнее тех заемщиков, которые не состояли в браке (не женат/не замужем, граждаский брак)

In [30]:
# получение данных через сводную таблицу
data_pivot_family = data.pivot_table(index='family_status', values='debt', aggfunc={'count', 'mean'})
data_pivot_family

count      mean
family_status                           
Не женат / не замужем   2796.0  0.097639
в разводе               1189.0  0.070648
вдовец / вдова           951.0  0.066246
гражданский брак        4134.0  0.093130
женат / замужем        12261.0  0.075606

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [31]:
# Категоризация заказчика

# Объявим функцию для присвоения категории уровня дохода
def income_categorize(income):
    try:
        if income < 30000:
            return 'A'
        elif income < 50000:
            return 'B'
        elif income < 200000:
            return 'C'
        elif income < 1000000:
            return 'D'
        else:
            return 'E'
    except:
        pass

# Создадим новый столбец для категории уровня дохода и заполним его значениями с помощью созданной выше функции    
data['income_category'] = data['total_income'].apply(income_categorize)

# Посмотрим на получившееся количество в группах
data.groupby('income_category')['debt'].count()

income_category
A       22
B      349
C    15921
D     5014
E       25
Name: debt, dtype: int64

In [32]:
# получение данных через сводную таблицу
data_pivot_income = data.pivot_table(index='income_category', values='debt', aggfunc={'count', 'mean'})
data_pivot_income

count      mean
income_category                   
A                   22.0  0.090909
B                  349.0  0.060172
C                15921.0  0.084982
D                 5014.0  0.070602
E                   25.0  0.080000

**Вывод:**
1. Категоризация заказчика:
Так как группы А, В и Е слишком малочисленны мы можем сравнить только две оставшиеся группы. Вывод из этой части исследования: заемщики с доходом от 50 до 200 тыс. менее благонадежны, чем заемщики с доходом от 200 тыс до 1 млн.

In [33]:
# Предлагаемая категоризация
# Объявим функцию для присвоения категории уровня дохода
def income_categorize(income):
    try:
        if income < 100000:
            return 'A'
        elif income < 150000:
            return 'B'
        elif income < 200000:
            return 'C'
        elif income < 250000:
            return 'D'
        else:
            return 'E'
    except:
        pass

# Создадим новый столбец для категории уровня дохода и заполним его значениями с помощью созданной выше функции    
data['income_category'] = data['total_income'].apply(income_categorize)

# Посмотрим на получившееся количество в группах
data.groupby('income_category')['debt'].count()

income_category
A    4444
B    7110
C    4738
D    2242
E    2797
Name: debt, dtype: int64

In [34]:
# Следующим шагом увидим % проблемных заемщиков в группе
ratio_by_income = data.groupby('income_category')['debt'].mean()
(ratio_by_income * 100).round(2).astype(float).astype(str) + '%'

income_category
A    7.97%
B    8.71%
C    8.51%
D    7.23%
E    6.94%
Name: debt, dtype: object

In [32]:
# получение данных через сводную таблицу
data_pivot_income = data.pivot_table(index='income_category', values='debt', aggfunc={'count', 'mean'})
data_pivot_income

count      mean
income_category                  
A                4444.0  0.079658
B                7110.0  0.087060
C                4738.0  0.085057
D                2242.0  0.072257
E                2797.0  0.069360

**Вывод:**
2. Предлагаемая категоризация:
Меньше всего проблем с возвратом кредитов у заемщиков с уровнем дохода более 250.000 (группа Е), и больше всего проблем у заемщиков с доходами в диапазоне от 100.000 до 200.000 (группы В и С)

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [33]:
# Посмотрим на распределение заемщиков по группам (по цели кредита)
data.groupby('purpose_category')['debt'].count()

purpose_category
операции с автомобилем       4279
операции с недвижимостью    10751
получение образования        3988
проведение свадьбы           2313
Name: debt, dtype: int64

In [34]:
# Посмотрим на распределение заемщиков по группам (по цели кредита)
ratio_by_purpose = data.groupby('purpose_category')['debt'].mean()
(ratio_by_purpose * 100).round(2).astype(float).astype(str) + '%'

purpose_category
операции с автомобилем      9.35%
операции с недвижимостью    7.26%
получение образования       9.25%
проведение свадьбы          7.91%
Name: debt, dtype: object

In [35]:
# получение данных через сводную таблицу
data_pivot_purpose = data.pivot_table(index='purpose_category', values='debt', aggfunc={'count', 'mean'})
data_pivot_purpose

count      mean
purpose_category                           
операции с автомобилем     4279.0  0.093480
операции с недвижимостью  10751.0  0.072551
получение образования      3988.0  0.092528
проведение свадьбы         2313.0  0.079118

**Вывод:**
Меньше проблем с возвратом кредитов у заемщиков с целями недвижимость и свадьба. К обоим решениям подходят очень ответственно и серьезно. Создание семьи - это самый ответственный шаг для человека, а "крыша над головой" это основа благосостояния.

Aвтомобиль это ТС повышенной опасности и возможными крупными дополнительными расходами (ремонт/обслуживание), а для заемщиков в категории "образование" одной из возможных причин возникновения проблем может стать сложность совмещения работы и образования и как следствие появление финансовых сложностей.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 
Заемщики могут не заполнять определенные поля в анкетах или онлайн-формах, по ошибке(или сознательно) пропуская важные сведения.
Некоторые поля могут быть необязательными для заполнения.
Могут возникать технические ошибки/сбои в процессе получения, сохранения и передачи информации от заемщиков.
Возможны ошибки в программном обеспечении или системных процессах, которые приводят к пропускам.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:*
Медианное значение позволяет минимизировать влияние экстремальных значений(когда некоторые значения сильно выделяются среди большинства). Например в штате из 5 человек четверо получают зарплату 5 тыс, а один 80 тыс - в среднем по 20 тыс, что не будет являться корректным описанием выборки. Можно либо поработать с данными до анализа и исключить слишком выдающиеся результаты либо взять медиану, которая в данном случае = 5 тыс. и более корректно покажет ситуацию для большинства в выбранной группе.

### Шаг 4: общий вывод.

Описание проекта
Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.


Описание данных
таблица состоящая из 12 столбцов и 21525 строк

children — количество детей в семье
days_employed — общий трудовой стаж в днях
dob_years — возраст клиента в годах
education — уровень образования клиента
education_id — идентификатор уровня образования
family_status — семейное положение
family_status_id — идентификатор семейного положения
gender — пол клиента
income_type — тип занятости
debt — имел ли задолженность по возврату кредитов
total_income — ежемесячный доход
purpose — цель получения кредита

Для проведения анализа были осуществлены следующие шаги:
*изучены предоставленые данные
*выявлены пропуски информации, ценность пропущенной информации и возможность ее восполнения
*проведена категоризация данных для целей анализа
*где возможно, данные были восполнены усредненными(медианными) значениями с учетом проведенной категоризации
*проведен анализ параметров/данных/статуса заемщиков и кредитов для выявления взаимосвязи с уровнем риска возникновения проблем по возврату

Результатом анализа являются ответы на вопросы:
Есть ли зависимость между количеством детей и возвратом кредита в срок?
Есть ли зависимость между семейным положением и возвратом кредита в срок?
Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
Как разные цели кредита влияют на его возврат в срок?

По результату проведенного исследования установлено, что:
Когда-либо состоявшие в браке/ либо состоящие сейчас, не имеющие детей, имеющие ежемесячный доход от 200 тыс до 1 млн. (либо при использовании предлагаемой категоризации от 200 тыс.) и взявшие кредит для покупки недвижимости - это самые благонадежные заемщики.
Не оформившие свои отношения официально (не женатыые/не замужем), имеющие двоих детей, доход от 50 до 200 тыс (либо при использовании предлагаемой категоризации от 100 до 200 тыс.) и взявшие кредит для покупки авто - это группа с наибольшим уровнем риска.


В качестве рекомендации:
рассмотреть другие уровни дохода для категоризации - для более корректной оценки влияния доходов на платежеспособность
рассмотреть сгруппированные данные сразу по нескольким предложенным проблемным вопросам - например сравнить платежеспособность заемщиков с самым высоким доходом с детьми и без детей. 



